File: java/src/com/company/datastructureimpl/DisjointSet.java
package com.company.datastructureimpl;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class DisjointSet {

    public Set<NodePath> nodeSet = new HashSet<>();

    public static void makeSet(Set<NodePath> set) {
        for (var node : set) {
            var ds = new DisjointSet();
            ds.nodeSet.add(node);
            node.disjointSet = ds;
        }
    }

    public static DisjointSet findSet(NodePath node) {
        return node.disjointSet;
    }

    public static void union(NodePath n1, NodePath n2) {
        if (n1.disjointSet == n2.disjointSet) return;

        if (n1.disjointSet.nodeSet.size() > n2.disjointSet.nodeSet.size()) {
            for (NodePath node : n2.disjointSet.nodeSet) {
                node.disjointSet = n1.disjointSet;
                n1.disjointSet.nodeSet.add(node);
            }

            return;
        }

        for (NodePath node : n1.disjointSet.nodeSet) {
            node.disjointSet = n2.disjointSet;
            n2.disjointSet.nodeSet.add(node);
        }
    }

    public static void main(String[] args) {
        var A = new NodePath("A");
        var B = new NodePath("B");
        var C = new NodePath("C");
        var D = new NodePath("D");
        var set = new HashSet<>(Set.of(A, B, C, D));
        DisjointSet.makeSet(set);

        union(A,B);
        union(A,D);
        System.out.println("end");
    }
}

class NodePath {
    public String name;
    public DisjointSet disjointSet;

    public NodePath(String name) {
        this.name = name;
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(name);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) return false;
        if (!(obj instanceof NodePath s)) return false;
        if (s.name == null || name == null) return false;
        return s.name.equals(name);
    }
}

File: kotlin/src/main/kotlin/algo/DisjointSet.kt
package algo

class WNode(var set: DisjointSet? = null, val name: String) {
    override fun toString(): String = name
}

class DisjointSet(var nodeList: MutableList<WNode> = mutableListOf())

fun makeSet(nodeList: MutableList<WNode>) {
    for (node in nodeList) {
        val set = DisjointSet()
        node.set = set
        node.set!!.nodeList.add(node)
    }
}

fun findSet(node: WNode): DisjointSet = node.set!!

fun union(node1: WNode, node2: WNode) {

    if (node1.set == node2.set) return

    val s1: DisjointSet = node1.set!!
    val s2: DisjointSet = node2.set!!

    if (s1.nodeList.size > s2.nodeList.size) {
        for (node in s2.nodeList) {
            node.set = s1
            s1.nodeList.add(node)
        }
        return
    }

    for (node in s1.nodeList) {
        node.set = s2
        s2.nodeList.add(node)
    }
}

fun printSet(set: DisjointSet) {
    for (node in set.nodeList) {
        print("${node.name},")
    }
}

// test

fun main() {
    val A = WNode(name = "A")
    val B = WNode(name = "B")
    val C = WNode(name = "C")
    val D = WNode(name = "D")

    val nodeList = mutableListOf(A, B, C, D)

    makeSet(nodeList)
    union(A, B)
    union(A, C)
    union(A, D)
    printSet(A.set!!)
}